src/Identity/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -44,11 +45,12 @@ public class ResourceOwnerPasswordValidator : BaseRequestValidator<ResourceOwner
         IPolicyService policyService,
         IDataProtectorTokenFactory<SsoEmail2faSessionTokenable> tokenDataFactory,
         IFeatureService featureService,
-        ISsoConfigRepository ssoConfigRepository)
+        ISsoConfigRepository ssoConfigRepository,
+        IDistributedCache distributedCache)
         : base(userManager, deviceRepository, deviceService, userService, eventService,
               organizationDuoWebTokenProvider, organizationRepository, organizationUserRepository,
               applicationCacheService, mailService, logger, currentContext, globalSettings, userRepository, policyService,
-              tokenDataFactory, featureService, ssoConfigRepository)
+              tokenDataFactory, featureService, ssoConfigRepository, distributedCache)
```

Comment:
❓ What scopes this cache to this use case of user-specific TOTP codes? Is this automatic and somehow scoped to the service? I don't have a lot of experience with `IDistributedCache`. My concern is another usage with email addresses could clash.

Reply:
It's a [singleton ](https://github.com/bitwarden/server/blob/d94a54516eafe2d86ca6288d19b54adf7b7bf284/src/SharedWeb/Utilities/ServiceCollectionExtensions.cs#L665)so this is an issue. I'll add information to the key to avoid any issues with the email being the Key.

Reply:
Added "TOTP_" prefix to the key to address the singleton issue.